### ISSUE TITLE
`--dry-run=server` support in admission webhook

### ISSUE THREAD

**Issue by hmlkao** [prometheus-operator/prometheus-operator#3704]
**What is missing?**
Admission webhook doesn't support `--dry-run=server`

**Why do we need it?**
To check Prometheus rules over `--dry-run=server` to be sure that Prometheus rule syntax is correct before deployment.

* Prometheus Operator version:
    `quay.io/coreos/prometheus-operator:v0.38.1`
* Kubectl version:
    `v1.19.4`

**Issue example**
```
cat invalid-prom-rule.yaml
---
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    prometheus: main
    role: alert-rules
  name: invalid-rule-dev-rules
spec:
  groups:
  - interval: 60s
    name: dev
    rules:
    - alert: InvalidPrometheusRule
      annotations:
        summary: This is invalid test alert
      expr: prometheus_build_info{version~=".*"}
      for: 10m
      labels:
        severity: "3"
```

```
$ kubectl apply -f invalid-prom-rule.yaml --dry-run=server
Error from server (BadRequest): error when creating "invalid-prom-rule.yaml": admission webhook "prometheusrulemutate.monitoring.coreos.com" does not support dry run
```

**Comment by simonpasquier** [comment]
Hmm I would expect that `--dry-run` is supported if the admission webhook is configured with `sideEffects: None` (according to https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#side-effects). Can you check if it is the case or not?

**Comment by hmlkao** [comment]
Honestly, I dindn't know about `sideEffeects` field so thanks for the link.
I changed this field to `sideEffects: None` (default is `Unknown`) as you mentioned but the behavior is still the same
```
$ kc get validatingwebhookconfigurations.admissionregistration.k8s.io prometheus-operator-admission -o yaml | grep sideEffects
  sideEffects: None
```
```
$ kubectl apply -f ~/Downloads/invalid-prom-rule.yaml --dry-run=server
Error from server (BadRequest): error when creating "/home/ondra/Downloads/invalid-prom-rule.yaml": admission webhook "prometheusrulemutate.monitoring.coreos.com" does not support dry run
```
I'm still convinced that it must be supported on the webhook side.

I use K8s `v1.16.14`, server-side dry run si [enabled by default from 1.13](https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/) so it should work.

**Comment by hmlkao** [comment]
I realized that the `expr` in example is valid PromQL request (and it will pass validation) so I have changed it to
```
      expr: prometheus_build_info{version~=".*"}
```
which is really invalid (correct is `version=~".*"`).

But that doesn't change fact that validation still not work even if field `sideEffect: None` is set...

**Comment by simonpasquier** [comment]
I did a test with v0.43.2 and I can't reproduce:

```
$ kubectl apply -f badrules.yaml --dry-run=server
The  "prometheusrules" is invalid: : 1:147: group "dev", rule 1, "InvalidPrometheusRule": could not parse expression: 1:30: parse error: unexpected character inside braces: '~'
```

Maybe it was broken before #3574? So I'd advise to update the operator's version.

**Comment by hmlkao** [comment]
Well, I finaly found time to check it again

# TL;DR
Can confirm that it was fixed in prometheus-operator v0.43.0 

# Long version
```
$ minikube version
minikube version: v1.16.0
commit: 9f1e482427589ff8451c4723b6ba53bb9742fbb1
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:51:19Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

```
$ minikube start --kubernetes-version=v1.20.0 --memory=4096 --extra-config=apiserver.authorization-mode=RBAC,Node
```

```
$ helm3 repo add prometheus-community https://prometheus-community.github.io/helm-charts
```

```
$ cat values.yaml
defaultRules:
  create: false

alertmanager:
  enabled: false

grafana:
  enabled: false

kubeApiServer:
  enabled: false

kubelet:
  enabled: false

kubeControllerManager:
  enabled: false

coreDns:
  enabled: false

kubeDns:
  enabled: false

kubeEtcd:
  enabled: false

kubeScheduler:
  enabled: false

kubeProxy:
  enabled: false

kubeStateMetrics:
  enabled: false

nodeExporter:
  enabled: false

prometheusOperator:
  enabled: true

prometheus:
  enabled: false
```

```
$ cat invalid-prom-rule.yaml
---
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  labels:
    prometheus: main
    role: alert-rules
  name: invalid-rule-rules
spec:
  groups:
  - name: invalid-prom-rule-1
    interval: 60s
    rules:
    - alert: InvalidPrometheusRule
      annotations:
        summary: This is invalid alert
      expr: prometheus_build_info{version~=".*"}
      for: 10m
      labels:
        severity: "3"
```

# Versions

## Chart 12.8.1 (prometheus-operator v0.44.0)

```
$ helm3 install kube-prometheus-stack prometheus-community/kube-prometheus-stack -f values.yaml --version v12.8.1
```

```
$ kubectl apply -f invalid-prom-rule.yaml --dry-run=server
The  "prometheusrules" is invalid: : 1:142: group "invalid-prom-rule-1", rule 1, "InvalidPrometheusRule": could not parse expression: 1:30: parse error: unexpected character inside braces: '~'
```

## Chart 11.0.0 (prometheus-operator v0.43.0)

```
$ helm3 install kube-prometheus-stack prometheus-community/kube-prometheus-stack -f values.yaml --version v11.0.0
```

```
$ kubectl apply -f invalid-prom-rule.yaml --dry-run=server
The  "prometheusrules" is invalid: : 1:158: group "invalid-prom-rule-1", rule 1, "InvalidPrometheusRule": could not parse expression: 1:30: parse error: unexpected character inside braces: '~'
```

## Chart 10.0.0 (prometheus-operator v0.42.1)

```
$ helm3 install kube-prometheus-stack prometheus-community/kube-prometheus-stack -f values.yaml --version v10.0.0
```

```
$ kubectl apply -f invalid-prom-rule.yaml --dry-run=server
Error from server (BadRequest): error when creating "invalid-prom-rule.yaml": admission webhook "prometheusrulemutate.monitoring.coreos.com" does not support dry run
```

**Comment by rubber-ant** [comment]
> Honestly, I dindn't know about `sideEffeects` field so thanks for the link. I changed this field to `sideEffects: None` (default is `Unknown`) as you mentioned but the behavior is still the same
> 
> ```
> $ kc get validatingwebhookconfigurations.admissionregistration.k8s.io prometheus-operator-admission -o yaml | grep sideEffects
>   sideEffects: None
> ```
> 
> ```
> $ kubectl apply -f ~/Downloads/invalid-prom-rule.yaml --dry-run=server
> Error from server (BadRequest): error when creating "/home/ondra/Downloads/invalid-prom-rule.yaml": admission webhook "prometheusrulemutate.monitoring.coreos.com" does not support dry run
> ```
> 
> I'm still convinced that it must be supported on the webhook side.
> 
> I use K8s `v1.16.14`, server-side dry run si [enabled by default from 1.13](https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/) so it should work.


@hmlkao  I followed your steps and it didn't work 
after I also changed the same way from `Unkown` to `None`  for:
```
kubectl get MutatingWebhookConfiguration  prometheus-operator-admission -oyaml | grep sideEffect
  sideEffects: None
```
and this solve my issue, thanks !


-----------

Error:
`Error from server (BadRequest): admission webhook "prometheusrulemutate.monitoring.coreos.com" does not support dry run`

K8s:
Client Version: v1.21.4
Server Version: v1.19.13-gke.1200

Version:
CHART 9.3.2   from  https://prometheus-community.github.io/helm-charts  (prometheus-operator	0.38.1)
